plugins/destination/s3/client/write.go

```diff
@@ -66,10 +72,37 @@ func (c *Client) WriteTableBatch(ctx context.Context, table *schema.Table, data
 	return nil
 }
 
-// sanitizeJSONKeys replaces all invalid characters in JSON keys with underscores.
-// It does the replacement in-place, modifying the original object. This is required
-// for compatibility with Athena.
-func sanitizeJSONKeys(obj any) {
+// sanitizeRecordJSONKeys replaces all invalid characters in JSON keys with underscores. This is required
+// for compatibility with Athena. It returns a new record, and the old one is released.
+func sanitizeRecordJSONKeys(mem memory.Allocator, record arrow.Record) (arrow.Record, error) {
+	newRecordBuilder := array.NewRecordBuilder(mem, record.Schema())
+	defer record.Release()
+	defer newRecordBuilder.Release()
+	for i := 0; i < int(record.NumCols()); i++ {
+		col := record.Column(i)
+		if arrow.TypeEqual(col.DataType(), types.NewJSONType()) {
+			for r := 0; r < int(record.NumRows()); r++ {
+				obj := col.GetOneForMarshal(r)
+				sanitizeJSONKeysForObject(obj)
+				newRecordBuilder.Field(i).(*types.JSONBuilder).Append(obj)
+			}
+			continue
+		}
+		for r := 0; r < int(record.NumRows()); r++ {
```

Comment:
I think there is no need to use the `Marshal/Unmarshall` for the unchanged columns. You can just use `array.NewRecord` and pass the columns. This way it will be "zero-copy". and just use the builder for the JSONarray not the whole record.

Reply:
Ah nice, will try that

Reply:
Cool, that works, thanks for the suggestion!